
<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Script Reference Guide &#8212; InPheRNo-ChIP 1.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css?v=eafc0fe6" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6" />
  <script src="_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=365ca57ee442770a23c6"></script>

    <script src="_static/documentation_options.js?v=61243dd2"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'usage';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Using InPheRNo-ChIP with Your Own Data" href="usage_own_data.html" />
    <link rel="prev" title="Installation Guide" href="installation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">InPheRNo-ChIP 1.2 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="introduction.html">
                        Introduction
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="installation.html">
                        Installation Guide
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        Script Reference Guide
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="usage_own_data.html">
                        Using InPheRNo-ChIP with Your Own Data
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          
 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">
 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="introduction.html">
                        Introduction
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="installation.html">
                        Installation Guide
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        Script Reference Guide
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="usage_own_data.html">
                        Using InPheRNo-ChIP with Your Own Data
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page">Script...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="script-reference-guide">
<h1>Script Reference Guide<a class="headerlink" href="#script-reference-guide" title="Link to this heading">#</a></h1>
<p>This guide provides detailed information on function parameters and code documentation for <strong>InPheRNo-ChIP</strong>.
The guide is divided into three main sections, corresponding to the three scripts of the <strong>InPheRNo-ChIP</strong> framework:</p>
<section id="inpherno-chip-step-1">
<h2>InPheRNo-ChIP Step 1<a class="headerlink" href="#inpherno-chip-step-1" title="Link to this heading">#</a></h2>
<p>The first step involves processing RNA-seq and ChIP-seq data to obtain 3 sets of pvalues.</p>
<p><strong>Basic Usage</strong>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>InPheC_Step1.py
</pre></div>
</div>
<p><strong>Advanced Usage</strong>:</p>
<p id="module-InPheC_Step1">This script is part of the InPheRNo-ChIP project and is responsible for process and analyze RNA-seq and ChIP-seq data.
It includes functionality for:</p>
<blockquote>
<div><ul class="simple">
<li><p>Parsing command-line arguments to configure paths and options for data processing.</p></li>
<li><p>Loading ChIP-seq and RNA-seq data from specified directories: ./Data/</p></li>
<li><p>Processing and filtering data to ensure consistency among gene and TF sets.</p></li>
<li><p>Performing Elastic Net regression on gene and TF expression data.</p></li>
</ul>
</div></blockquote>
<p>Usage:</p>
<blockquote>
<div><ul class="simple">
<li><p>Run the script from the command line “python InPheC_Step1.py” directly.</p></li>
<li><p>For detailed argument options, use the help option: <cite>python InPheC_Step1.py –help</cite></p></li>
</ul>
</div></blockquote>
<p>Note: Ensure that all dependencies are installed and the Python environment is correctly set up for running this script.</p>
<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step1.DataLoader">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step1.</span></span><span class="sig-name descname"><span class="pre">DataLoader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chip_seq_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_seq_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#DataLoader"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.DataLoader" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Loads and prepares RNA-seq and ChIP-seq data from specified directories.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chip_seq_dir</strong> (<em>str</em>) – Path to the ChIP-seq data directory.</p></li>
<li><p><strong>rna_seq_dir</strong> (<em>str</em>) – Path to the RNA-seq data directory.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step1.DataLoader.load_chip_data">
<span class="sig-name descname"><span class="pre">load_chip_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lineage_of_interest</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#DataLoader.load_chip_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.DataLoader.load_chip_data" title="Link to this definition">#</a></dt>
<dd><p>Loads and merges ChIP-seq data files for specified lineages.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lineage_of_interest</strong> (<em>list</em>) – Lineages to include in the analysis.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Combined DataFrame of ChIP-seq data across specified lineages.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step1.DataLoader.load_rna_data">
<span class="sig-name descname"><span class="pre">load_rna_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f_exp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_gp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#DataLoader.load_rna_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.DataLoader.load_rna_data" title="Link to this definition">#</a></dt>
<dd><p>Loads RNA-seq expression data and associated gene-phenotype p-values from specified files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f_exp</strong> (<em>str</em>) – Filename of the RNA-seq expression data file.</p></li>
<li><p><strong>f_gp</strong> (<em>str</em>) – Filename of the gene-phenotype p-values data file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing expression data and gene-phenotype p-values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pandas.DataFrame, pandas.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step1.DataLoader.load_tf_list">
<span class="sig-name descname"><span class="pre">load_tf_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f_tf</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#DataLoader.load_tf_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.DataLoader.load_tf_list" title="Link to this definition">#</a></dt>
<dd><p>Loads a list of transcription factors from a specified CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>f_tf</strong> (<em>str</em>) – Path to the CSV file containing the list of transcription factors.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing the list of transcription factors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step1.DataProcessor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step1.</span></span><span class="sig-name descname"><span class="pre">DataProcessor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chip_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">GEx_gene_tf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_phenotype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hm_tf_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chip_pvalue_of_interest</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#DataProcessor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.DataProcessor" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Processes and filters RNA-seq and ChIP-seq data to ensure consistency among gene and TF sets.
This class is crucial for preparing the data for subsequent analysis, including filtering based on shared genes and TFs.</p>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step1.DataProcessor.filter_data">
<span class="sig-name descname"><span class="pre">filter_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#DataProcessor.filter_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.DataProcessor.filter_data" title="Link to this definition">#</a></dt>
<dd><p>Filters RNA-seq and ChIP-seq data to ensure consistency among gene and transcription factor (TF) sets,
and prepares the data for subsequent analysis steps. This method processes the data by:
- Identifying common genes and TFs across different datasets.
- Filtering ChIP-seq data to keep rows that have genes and TFs present in both lineages.
- Filtering to keep the minimal p-value across peaks for each (TF, gene, lineage) combination.
- Preparing expression data matrices for genes and TFs for Elastic Net regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple containing:
- pandas.DataFrame: Filtered ChIP-seq data with minimal p-values across peaks, reset index.
- pandas.DataFrame: Gene expression data for shared genes.
- pandas.DataFrame: TF expression data for shared TFs.
- pandas.DataFrame: Gene-phenotype data for shared genes.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step1.ElasticNetProcessor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step1.</span></span><span class="sig-name descname"><span class="pre">ElasticNetProcessor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expr_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expr_tf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_num_coefs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#ElasticNetProcessor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.ElasticNetProcessor" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Conducts Elastic Net regression to identify relationships between gene expressions and transcription factors
within the InPheRNo-ChIP project. It dynamically adjusts model parameters to optimize fit and manage convergence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>expr_gene</strong> (<em>pandas.DataFrame</em>) – DataFrame containing gene expression data.</p></li>
<li><p><strong>expr_tf</strong> (<em>pandas.DataFrame</em>) – DataFrame containing TF expression data.</p></li>
<li><p><strong>max_num_coefs</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum allowed non-zero coefficients in the model, provides a control over model complexity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step1.ElasticNetProcessor.perform_elastic_net">
<span class="sig-name descname"><span class="pre">perform_elastic_net</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#ElasticNetProcessor.perform_elastic_net"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.ElasticNetProcessor.perform_elastic_net" title="Link to this definition">#</a></dt>
<dd><p>Performs Elastic Net regression across all genes, using transcription factor expressions to predict gene activity.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame with p-values indicating the significance of associations between genes and TFs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Note<span class="colon">:</span></dt>
<dd class="field-odd"><p>This function is adapted from the InPheRNo paper with updated eps and n_alphas values to improve model performance and stability.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step1.Step1ArgumentParser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step1.</span></span><span class="sig-name descname"><span class="pre">Step1ArgumentParser</span></span><a class="reference internal" href="_modules/InPheC_Step1.html#Step1ArgumentParser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.Step1ArgumentParser" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Handles command-line arguments for the InPheRNo-ChIP project, setting up and parsing them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parser</strong> (<em>argparse.ArgumentParser</em>) – Configures the command-line arguments.</p></li>
<li><p><strong>args</strong> (<em>argparse.Namespace</em>) – Stores the values of parsed arguments.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step1.Step1ArgumentParser.parse_args">
<span class="sig-name descname"><span class="pre">parse_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#Step1ArgumentParser.parse_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.Step1ArgumentParser.parse_args" title="Link to this definition">#</a></dt>
<dd><p>Parses the command-line arguments.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="InPheC_Step1.save_output_files">
<span class="sig-prename descclassname"><span class="pre">InPheC_Step1.</span></span><span class="sig-name descname"><span class="pre">save_output_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chip_gt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_gt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_gp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#save_output_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.save_output_files" title="Link to this definition">#</a></dt>
<dd><p>Saves processed data to specified output files as part of the InPheRNo-ChIP project. This function handles the
creation of output directories and manages the naming conventions for output files based on the provided command-line arguments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>argparse.Namespace</em>) – Parsed command-line arguments containing output file paths and names.</p></li>
<li><p><strong>chip_gt</strong> (<em>pandas.DataFrame</em>) – Processed ChIP-seq data to be saved.</p></li>
<li><p><strong>rna_gt</strong> (<em>pandas.DataFrame</em>) – Processed RNA-seq gene-TF data to be saved.</p></li>
<li><p><strong>rna_gp</strong> (<em>pandas.DataFrame</em>) – Processed RNA-seq gene-phenotype data to be saved.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="InPheC_Step1.step1_main">
<span class="sig-prename descclassname"><span class="pre">InPheC_Step1.</span></span><span class="sig-name descname"><span class="pre">step1_main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step1.html#step1_main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step1.step1_main" title="Link to this definition">#</a></dt>
<dd><p>Main function for the first step in the InPheRNo-ChIP project pipeline. Orchestrates the execution of data loading,
processing, Elastic Net regression, and saving the output files. This function acts as the entry point when running the script.</p>
<p>This function follows these steps:
- 1. Parses command-line arguments.
- 2. Loads data from specified directories.
- 3. Processes data to align gene and TF datasets.
- 4. Performs Elastic Net regression to analyze gene-TF relationships.
- 5. Saves the processed data to output files.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>SystemExit</strong> – If the dimensions of gene-phenotype data and results from Elastic Net regression do not match, indicating a potential issue in data alignment or processing.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="inpherno-chip-step-2">
<h2>InPheRNo-ChIP Step 2<a class="headerlink" href="#inpherno-chip-step-2" title="Link to this heading">#</a></h2>
<p>The second step involves running the probabilistic graphical model (PGM).</p>
<p><strong>Basic Usage</strong>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>InPheC_Step2.py
</pre></div>
</div>
<p><strong>Advanced Usage</strong>:</p>
<p id="module-InPheC_Step2">InPheC_Step2.py</p>
<p>This script is part of the InPheRNo-ChIP project and is responsible for:</p>
<blockquote>
<div><ul class="simple">
<li><p>Estimating the parameters of the model.</p></li>
<li><p>Running the probabilistic graphical model (PGM).</p></li>
<li><p>Incorporating enhancements in model Qij to account for inter-lineage dependencies and maintaining numerical stability.</p></li>
</ul>
</div></blockquote>
<p>Usage:</p>
<blockquote>
<div><ul class="simple">
<li><p>Run the script from the command line “python InPheC_Step2.py” directly.</p></li>
<li><p>For detailed argument options, use the help option: <cite>python InPheC_Step2.py –help</cite></p></li>
</ul>
</div></blockquote>
<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step2.CorePGM">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step2.</span></span><span class="sig-name descname"><span class="pre">CorePGM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in2pgm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pgm_param</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Prior_T</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#CorePGM"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.CorePGM" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class sets up the probabilistic graphical model (PGM) configurations, manages batch processing of genes, and handles the execution of the model for each gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>in2pgm</strong> (<em>dict</em>) – Input data and configurations for setting up the PGM.</p></li>
<li><p><strong>pgm_param</strong> (<em>dict</em>) – Parameters to control the PGM execution, such as number of iterations and chains.</p></li>
<li><p><strong>Prior_T</strong> (<em>float</em>) – Prior probability for the Bernoulli distribution of Tij.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.CorePGM.logsumexp">
<span class="sig-name descname"><span class="pre">logsumexp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">logp1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logp2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#CorePGM.logsumexp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.CorePGM.logsumexp" title="Link to this definition">#</a></dt>
<dd><p>Computes the log-sum-exp of two log probabilities for numerical stability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>logp1</strong> (<em>float</em>) – Log probability 1.</p></li>
<li><p><strong>logp2</strong> (<em>float</em>) – Log probability 2.</p></li>
<li><p><strong>r</strong> (<em>float</em>) – Mixing ratio.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Result of log-sum-exp calculation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.CorePGM.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">out_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_chain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_batch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#CorePGM.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.CorePGM.run" title="Link to this definition">#</a></dt>
<dd><p>Processes genes in batches, executing the PGM for each batch and managing the output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out_dir</strong> (<em>str</em>) – Directory to save the output files.</p></li>
<li><p><strong>n_chain</strong> (<em>int</em>) – Number of chains for MCMC sampling.</p></li>
<li><p><strong>start_batch</strong> (<em>int</em>) – Starting index of the batch processing.</p></li>
<li><p><strong>batch_size</strong> (<em>int</em>) – Number of genes processed in each batch.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.CorePGM.run_model">
<span class="sig-name descname"><span class="pre">run_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">one_gene</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#CorePGM.run_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.CorePGM.run_model" title="Link to this definition">#</a></dt>
<dd><p>Executes the PGM for a single gene based on the provided gene data and model parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>one_gene</strong> (<em>pandas.DataFrame</em>) – Data for one gene including its interactions and expressions.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Trace of the model after sampling.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pymc.backends.base.MultiTrace</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.CorePGM.setup_model">
<span class="sig-name descname"><span class="pre">setup_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_TF</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_lineages</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a_gp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#CorePGM.setup_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.CorePGM.setup_model" title="Link to this definition">#</a></dt>
<dd><p>Configures the PGM model using the specified parameters and distributions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_TF</strong> (<em>int</em>) – Number of transcription factors.</p></li>
<li><p><strong>n_lineages</strong> (<em>int</em>) – Number of lineages, default to 1, as we are considering the minimum p-value across lineages.</p></li>
<li><p><strong>a_gp</strong> (<em>float</em>) – Estimated parameter ‘a_prime’ for the Beta distribution in phenotype modeling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Configured PGM model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pymc.Model</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step2.PGMInputPreparer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step2.</span></span><span class="sig-name descname"><span class="pre">PGMInputPreparer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">out_tmp_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_tg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_gene_pheno</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chip_tg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_gene_pheno_all</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#PGMInputPreparer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.PGMInputPreparer" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Prepares and processes input data for the probabilistic graphical model (PGM) as part of the InPheRNo-ChIP project.
This includes data masking, estimating parameters, and adding control variables to ensure robust model performance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out_tmp_dir</strong> (<em>str</em>) – Directory path for saving intermediate outputs.</p></li>
<li><p><strong>rna_tg</strong> (<em>pandas.DataFrame</em>) – DataFrame containing RNA-seq gene-TF associations.</p></li>
<li><p><strong>rna_gene_pheno</strong> (<em>pandas.DataFrame</em>) – DataFrame containing gene-phenotype associations.</p></li>
<li><p><strong>chip_tg</strong> (<em>pandas.DataFrame</em>) – DataFrame containing ChIP-seq gene-TF associations.</p></li>
<li><p><strong>rna_gene_pheno_all</strong> (<em>pandas.DataFrame</em>) – DataFrame containing gene-phenotype associations for all genes.</p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step2.PGMInputPreparer.beta_unif_gen">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">beta_unif_gen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">momtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xtol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-14</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">badvalue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">longname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shapes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#PGMInputPreparer.beta_unif_gen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.PGMInputPreparer.beta_unif_gen" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">rv_continuous</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.PGMInputPreparer.check_modality_per_gene">
<span class="sig-name descname"><span class="pre">check_modality_per_gene</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#PGMInputPreparer.check_modality_per_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.PGMInputPreparer.check_modality_per_gene" title="Link to this definition">#</a></dt>
<dd><p>Ensures that each gene meets specific criteria in both RNA-seq and ChIP-seq data for inclusion in the PGM.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrames filtered based on modality checks.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple of pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.PGMInputPreparer.combine_and_add_controls">
<span class="sig-name descname"><span class="pre">combine_and_add_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flt_rna_tg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flt_chip_tg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flt_rna_gp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#PGMInputPreparer.combine_and_add_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.PGMInputPreparer.combine_and_add_controls" title="Link to this definition">#</a></dt>
<dd><p>Merges datasets and adds control entries to prepare for PGM analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flt_rna_tg</strong> (<em>pandas.DataFrame</em>) – Filtered RNA-seq gene-TF associations.</p></li>
<li><p><strong>flt_chip_tg</strong> (<em>pandas.DataFrame</em>) – Filtered ChIP-seq gene-TF associations.</p></li>
<li><p><strong>flt_rna_gp</strong> (<em>pandas.DataFrame</em>) – Filtered gene-phenotype associations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Combined DataFrame with control entries added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.PGMInputPreparer.estimate_a_prime">
<span class="sig-name descname"><span class="pre">estimate_a_prime</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Pj_arr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#PGMInputPreparer.estimate_a_prime"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.PGMInputPreparer.estimate_a_prime" title="Link to this definition">#</a></dt>
<dd><p>Estimates the ‘a_prime’ parameter for the PGM using a hybrid beta-uniform distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>Pj_arr</strong> (<em>numpy.ndarray</em>) – Array of p-values for estimating the distribution parameter.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Estimated ‘a_prime’ parameter.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step2.Step1DataLoader">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step2.</span></span><span class="sig-name descname"><span class="pre">Step1DataLoader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#Step1DataLoader"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.Step1DataLoader" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Handles the loading of data from outputs of the first step in the InPheRNo-ChIP project pipeline as well as RNA-seq data.</p>
<p>This class is tasked with ensuring data integrity and format correctness as it prepares the data for subsequent processing stages.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>args</strong> (<em>argparse.Namespace</em>) – Parsed command-line arguments containing paths and file-related settings.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.Step1DataLoader.load_csv">
<span class="sig-name descname"><span class="pre">load_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#Step1DataLoader.load_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.Step1DataLoader.load_csv" title="Link to this definition">#</a></dt>
<dd><p>Loads a CSV file into a pandas DataFrame, using a specified delimiter.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.Step1DataLoader.load_data">
<span class="sig-name descname"><span class="pre">load_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#Step1DataLoader.load_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.Step1DataLoader.load_data" title="Link to this definition">#</a></dt>
<dd><p>Loads necessary data files from specified directories for further processing, including gene-TF associations,
gene-phenotype associations, and ChIP-seq data. Validates the loading process by printing the shapes of loaded datasets.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple containing DataFrames for gene-phenotype associations for all genes, gene-phenotype associations for the set of interest,
gene-TF associations, and ChIP-seq gene-TF associations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step2.Step2ArgumentParser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step2.</span></span><span class="sig-name descname"><span class="pre">Step2ArgumentParser</span></span><a class="reference internal" href="_modules/InPheC_Step2.html#Step2ArgumentParser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.Step2ArgumentParser" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Parses and handles command-line arguments for the step2_main.py script, ensuring proper validation and configuration
of input and output paths and file-related options for RNA-seq and ChIP-seq data processing.</p>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.Step2ArgumentParser.get_args">
<span class="sig-name descname"><span class="pre">get_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#Step2ArgumentParser.get_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.Step2ArgumentParser.get_args" title="Link to this definition">#</a></dt>
<dd><p>Returns the validated command-line arguments.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Parsed command-line arguments.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>argparse.Namespace</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step2.Step2ArgumentParser.validate_args">
<span class="sig-name descname"><span class="pre">validate_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#Step2ArgumentParser.validate_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.Step2ArgumentParser.validate_args" title="Link to this definition">#</a></dt>
<dd><p>Validates the existence of directories specified in command-line arguments and creates the output directory if it does not exist.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="InPheC_Step2.step2_main">
<span class="sig-prename descclassname"><span class="pre">InPheC_Step2.</span></span><span class="sig-name descname"><span class="pre">step2_main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step2.html#step2_main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step2.step2_main" title="Link to this definition">#</a></dt>
<dd><p>Entry point for executing the step 2 of the InPheRNo-ChIP project’s pipeline. This function orchestrates
the loading of data, preprocessing, and execution of the probabilistic graphical model (PGM) for gene
regulatory network (GRN) inference.</p>
<p>Steps:
- Parses command-line arguments for configuration settings.
- Loads data produced by step 1, including gene-TF and gene-phenotype associations.
- Prepares input for the PGM, including data cleaning, modality checks, and control integration.
- Configures and runs the PGM to infer regulatory networks.</p>
<p>The function manages file directories, initializes data preparation, sets PGM parameters, and
handles the batch processing of genes, ensuring all outputs are appropriately saved and managed.</p>
</dd></dl>

</section>
<section id="inpherno-chip-step-3">
<h2>InPheRNo-ChIP Step 3<a class="headerlink" href="#inpherno-chip-step-3" title="Link to this heading">#</a></h2>
<p>The final step combines and processes the output from the previous steps.</p>
<p><strong>Basic Usage</strong>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>InPheC_Step3.py
</pre></div>
</div>
<p><strong>Advanced Usage</strong>:</p>
<p id="module-InPheC_Step3">InPheC_Step3.py</p>
<p>This script is an integral part of the InPheRNo-ChIP project’s step 3 process.</p>
<p>Key Operations:</p>
<blockquote>
<div><ul class="simple">
<li><p>Combines multiple batches</p></li>
<li><p>Averages posterior probabilities across different chains from step 2’s PGM outputs.</p></li>
<li><p>Applies gene-wise filters to the averaged data.</p></li>
<li><p>Performs min-max normalization and form the final phenotype-relevant GRN.</p></li>
</ul>
</div></blockquote>
<p>Usage:</p>
<blockquote>
<div><ul class="simple">
<li><p>Run the script from the command line “python InPheC_Step3.py” directly.</p></li>
<li><p>For detailed argument options, use the help option: <cite>python InPheC_Step3.py –help</cite></p></li>
</ul>
</div></blockquote>
<p>Note:
Ensure the Python environment is properly set up and all dependencies are installed before running this script.</p>
<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step3.DataProcessor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">DataProcessor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step3.html#DataProcessor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.DataProcessor" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Processes and analyzes the combined posterior data from probabilistic graphical models.</p>
<p>This class loads and processes the in2pgm and posterior data, combines them, and applies filtering and normalization
to produce a final gene regulatory network (GRN).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fn1</strong> (<em>str</em>) – Path to the first input file containing in2pgm data.</p></li>
<li><p><strong>fn2</strong> (<em>str</em>) – Path to the second input file containing posterior data.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.DataProcessor.combine_in2pgm_outposterior">
<span class="sig-name descname"><span class="pre">combine_in2pgm_outposterior</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/InPheC_Step3.html#DataProcessor.combine_in2pgm_outposterior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.DataProcessor.combine_in2pgm_outposterior" title="Link to this definition">#</a></dt>
<dd><p>Combines the in2pgm and posterior data into a single DataFrame.</p>
<p>This function merges the two datasets on the TF and Gene columns, effectively creating an edge list for the GRN.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Combined DataFrame with the in2pgm and posterior data.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.DataProcessor.form_final_grn">
<span class="sig-name descname"><span class="pre">form_final_grn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col_raw_score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Tp_val</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/InPheC_Step3.html#DataProcessor.form_final_grn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.DataProcessor.form_final_grn" title="Link to this definition">#</a></dt>
<dd><p>Forms the final gene regulatory network (GRN) based on normalized and filtered data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col_raw_score</strong> (<em>str</em>) – Column name of the raw scores used for final processing.</p></li>
<li><p><strong>Tp_val</strong> (<em>float</em>) – Threshold prior value used in forming the GRN.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing the final GRN DataFrame and gene-specific thresholds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="InPheC_Step3.FileNotFound">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">FileNotFound</span></span><a class="reference internal" href="_modules/InPheC_Step3.html#FileNotFound"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.FileNotFound" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Custom exception for handling file-not-found errors.</p>
<p>This exception is raised when a required file is not found in the specified directory,
extending the base Exception class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>message</strong> (<em>str</em>) – The error message to display.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step3.ParameterExtractor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">ParameterExtractor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step3.html#ParameterExtractor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.ParameterExtractor" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Extracts parameters from file names in the specified directory.</p>
<p>Designed to read file names in a given directory and extract key parameters such as T_prior, n_c, n_i, and n_t, which are essential for processing steps in the pipeline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>in_dir</strong> (<em>str</em>) – Input directory where files are located.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing extracted parameters with keys like ‘sample_fn’, ‘T_prior’, ‘n_c’, ‘n_i’, and ‘n_t’.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, Union[str, float]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#InPheC_Step3.FileNotFound" title="InPheC_Step3.FileNotFound"><strong>FileNotFound</strong></a> – If no relevant pickle file is found in the directory.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.ParameterExtractor.extract">
<span class="sig-name descname"><span class="pre">extract</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/InPheC_Step3.html#ParameterExtractor.extract"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.ParameterExtractor.extract" title="Link to this definition">#</a></dt>
<dd><p>Extracts parameters from the filenames in the input directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary containing the extracted parameters.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict[str, Union[str, float]]</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#InPheC_Step3.FileNotFound" title="InPheC_Step3.FileNotFound"><strong>FileNotFound</strong></a> – If no relevant pickle file is found in the directory.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step3.PosteriorCombiner">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">PosteriorCombiner</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_batches</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_burn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">common_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step3.html#PosteriorCombiner"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.PosteriorCombiner" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Combines posterior distributions from multiple batches and chains.</p>
<p>Processes each batch, extracting and combining posterior distributions into a single DataFrame from pickle files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>in_dir</strong> (<em>str</em>) – Input directory containing the pickle files.</p></li>
<li><p><strong>n_batches</strong> (<em>int</em>) – Number of batches to process.</p></li>
<li><p><strong>n_burn</strong> (<em>int</em>) – Number of burn-in iterations.</p></li>
<li><p><strong>common_params</strong> (<em>dict</em>) – Common parameters used across the batches.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing paths to the combined posterior file and the in2pgm data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.PosteriorCombiner.combine">
<span class="sig-name descname"><span class="pre">combine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">out_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step3.html#PosteriorCombiner.combine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.PosteriorCombiner.combine" title="Link to this definition">#</a></dt>
<dd><p>Combines the posteriors and saves the results to specified output files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>out_dir</strong> (<em>str</em>) – Directory where the output files will be saved.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple of paths to the combined in2pgm and posterior files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[str, str]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step3.Step3ArgumentParser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">Step3ArgumentParser</span></span><a class="reference internal" href="_modules/InPheC_Step3.html#Step3ArgumentParser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.Step3ArgumentParser" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Parses command-line arguments for the script.</p>
<p>This class configures and retrieves command-line options, returning the parsed arguments as an argparse.Namespace object.
It sets default values for input and output directories, the number of batches, and the number of burn-in iterations for posterior processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An argparse.Namespace object containing all the command-line arguments.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>argparse.Namespace</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<p>args = ArgumentParser.parse()
print(args.in_dir)  # Prints the input directory specified in the command-line</p>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.Step3ArgumentParser.parse">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Namespace</span></span></span><a class="reference internal" href="_modules/InPheC_Step3.html#Step3ArgumentParser.parse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.Step3ArgumentParser.parse" title="Link to this definition">#</a></dt>
<dd><p>Sets up the command-line arguments and parses them.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Parsed arguments from the command line.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>argparse.Namespace</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="InPheC_Step3.Utils">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">Utils</span></span><a class="reference internal" href="_modules/InPheC_Step3.html#Utils"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.Utils" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.Utils.check_pkl_file_count">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">check_pkl_file_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expected_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step3.html#Utils.check_pkl_file_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.Utils.check_pkl_file_count" title="Link to this definition">#</a></dt>
<dd><p>Checks if the number of .pkl files in the input directory matches the expected number.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>in_dir</strong> (<em>str</em>) – Directory to check for .pkl files.</p></li>
<li><p><strong>expected_count</strong> (<em>int</em>) – The number of .pkl files expected in the directory.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the number of .pkl files does not match the expected count.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.Utils.configure_logging">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">configure_logging</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">log_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/InPheC_Step3.html#Utils.configure_logging"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.Utils.configure_logging" title="Link to this definition">#</a></dt>
<dd><p>Configures the logging system for the script, directing log output to a file in a specified directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>log_dir</strong> (<em>str</em>) – Directory where the log file will be stored.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="InPheC_Step3.Utils.print_python_details">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">print_python_details</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/InPheC_Step3.html#Utils.print_python_details"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.Utils.print_python_details" title="Link to this definition">#</a></dt>
<dd><p>Prints detailed information about the Python environment running the script.
Outputs various Python-related system information, including version, compiler, and other system details.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="InPheC_Step3.step3_main">
<span class="sig-prename descclassname"><span class="pre">InPheC_Step3.</span></span><span class="sig-name descname"><span class="pre">step3_main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/InPheC_Step3.html#step3_main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#InPheC_Step3.step3_main" title="Link to this definition">#</a></dt>
<dd><p>Executes the main functionality of the Step 3 script in the InPheRNo-ChIP project. This function manages the workflow
that combines and processes posterior distributions from probabilistic graphical models (PGMs) generated in Step 2.</p>
<p>The main steps include:
- Configuring logging for debugging and information tracking.
- Parsing command line arguments to get directories and processing parameters.
- Checking the count of .pkl files to ensure all expected files are present.
- Creating necessary directories for outputs.
- Extracting and logging parameters from file names.
- Checking for and combining existing processed data or processing new data.
- Merging, filtering, and normalizing data to form the final Gene Regulatory Network (GRN).
- Saving the final outputs in an Excel file with multiple sheets for easy review.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files are not found in the specified directories.</p></li>
<li><p><strong>ValueError</strong> – If processing parameters are incorrect or if files are empty.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="faqs">
<h2>FAQs<a class="headerlink" href="#faqs" title="Link to this heading">#</a></h2>
<p>Include a Frequently Asked Questions section to cover common queries:</p>
<ol class="arabic">
<li><p><strong>Question 1</strong>: <em>can’t wait for the 1st question you have for InPheC!</em></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<p>Answer 1:…</p>
</li>
<li><p><strong>Question 2</strong>:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<p>Answer 2:…</p>
</li>
</ol>
</section>
<section id="support-and-contribution">
<h2>Support and Contribution<a class="headerlink" href="#support-and-contribution" title="Link to this heading">#</a></h2>
<p>We are committed to providing support for <strong>InPheRNo-ChIP</strong> users and actively encourage contributions to the project.</p>
<section id="getting-support">
<h3>Getting Support<a class="headerlink" href="#getting-support" title="Link to this heading">#</a></h3>
<p>If you encounter issues or have questions while using <strong>InPheRNo-ChIP</strong>, there are several ways to get support:</p>
<ul class="simple">
<li><p><strong>Email the Author</strong>: Feel free to reach out to the author of the paper associated with <strong>InPheRNo-ChIP</strong>. The author’s email can typically be found in the corresponding research paper or on the project’s main webpage.</p></li>
<li><p><strong>GitHub Issues</strong>: For technical problems or bugs, you can open an issue on the GitHub repository.</p></li>
</ul>
</section>
<section id="contributing-to-inpherno-chip">
<h3>Contributing to InPheRNo-ChIP<a class="headerlink" href="#contributing-to-inpherno-chip" title="Link to this heading">#</a></h3>
<p>Your contributions to <strong>InPheRNo-ChIP</strong> are highly valued. There are various ways you can contribute to the project:</p>
<ul class="simple">
<li><p><strong>Reporting Bugs</strong>: If you find a bug, please report it by opening an issue on our GitHub repository. Provide as much detail as possible to help us understand and address the issue.</p></li>
<li><p><strong>Feature Requests</strong>: Have ideas for new features or improvements? We’d love to hear them! Please file a feature request on our Github Issue Page.</p></li>
<li><p><strong>Code Contributions</strong>: If you’re interested in contributing code, feel free to fork the repository and submit a pull request with your changes. For major changes, please open an issue first to discuss what you would like to change.</p></li>
</ul>
<p>We appreciate all forms of feedback and contributions to make <strong>InPheRNo-ChIP</strong> better for everyone!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This guide assumes a basic understanding of python and pymc. If you are new to these, consider reading <a class="reference external" href="https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/pymc_overview.html">PyMC documentation</a></p>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="installation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Installation Guide</p>
      </div>
    </a>
    <a class="right-next"
       href="usage_own_data.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Using InPheRNo-ChIP with Your Own Data</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inpherno-chip-step-1">InPheRNo-ChIP Step 1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.DataLoader"><code class="docutils literal notranslate"><span class="pre">DataLoader</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.DataLoader.load_chip_data"><code class="docutils literal notranslate"><span class="pre">DataLoader.load_chip_data()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.DataLoader.load_rna_data"><code class="docutils literal notranslate"><span class="pre">DataLoader.load_rna_data()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.DataLoader.load_tf_list"><code class="docutils literal notranslate"><span class="pre">DataLoader.load_tf_list()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.DataProcessor"><code class="docutils literal notranslate"><span class="pre">DataProcessor</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.DataProcessor.filter_data"><code class="docutils literal notranslate"><span class="pre">DataProcessor.filter_data()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.ElasticNetProcessor"><code class="docutils literal notranslate"><span class="pre">ElasticNetProcessor</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.ElasticNetProcessor.perform_elastic_net"><code class="docutils literal notranslate"><span class="pre">ElasticNetProcessor.perform_elastic_net()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.Step1ArgumentParser"><code class="docutils literal notranslate"><span class="pre">Step1ArgumentParser</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.Step1ArgumentParser.parse_args"><code class="docutils literal notranslate"><span class="pre">Step1ArgumentParser.parse_args()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.save_output_files"><code class="docutils literal notranslate"><span class="pre">save_output_files()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step1.step1_main"><code class="docutils literal notranslate"><span class="pre">step1_main()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inpherno-chip-step-2">InPheRNo-ChIP Step 2</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.CorePGM"><code class="docutils literal notranslate"><span class="pre">CorePGM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.CorePGM.logsumexp"><code class="docutils literal notranslate"><span class="pre">CorePGM.logsumexp()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.CorePGM.run"><code class="docutils literal notranslate"><span class="pre">CorePGM.run()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.CorePGM.run_model"><code class="docutils literal notranslate"><span class="pre">CorePGM.run_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.CorePGM.setup_model"><code class="docutils literal notranslate"><span class="pre">CorePGM.setup_model()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.PGMInputPreparer"><code class="docutils literal notranslate"><span class="pre">PGMInputPreparer</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.PGMInputPreparer.beta_unif_gen"><code class="docutils literal notranslate"><span class="pre">PGMInputPreparer.beta_unif_gen</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.PGMInputPreparer.check_modality_per_gene"><code class="docutils literal notranslate"><span class="pre">PGMInputPreparer.check_modality_per_gene()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.PGMInputPreparer.combine_and_add_controls"><code class="docutils literal notranslate"><span class="pre">PGMInputPreparer.combine_and_add_controls()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.PGMInputPreparer.estimate_a_prime"><code class="docutils literal notranslate"><span class="pre">PGMInputPreparer.estimate_a_prime()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.Step1DataLoader"><code class="docutils literal notranslate"><span class="pre">Step1DataLoader</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.Step1DataLoader.load_csv"><code class="docutils literal notranslate"><span class="pre">Step1DataLoader.load_csv()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.Step1DataLoader.load_data"><code class="docutils literal notranslate"><span class="pre">Step1DataLoader.load_data()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.Step2ArgumentParser"><code class="docutils literal notranslate"><span class="pre">Step2ArgumentParser</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.Step2ArgumentParser.get_args"><code class="docutils literal notranslate"><span class="pre">Step2ArgumentParser.get_args()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.Step2ArgumentParser.validate_args"><code class="docutils literal notranslate"><span class="pre">Step2ArgumentParser.validate_args()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step2.step2_main"><code class="docutils literal notranslate"><span class="pre">step2_main()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inpherno-chip-step-3">InPheRNo-ChIP Step 3</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.DataProcessor"><code class="docutils literal notranslate"><span class="pre">DataProcessor</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.DataProcessor.combine_in2pgm_outposterior"><code class="docutils literal notranslate"><span class="pre">DataProcessor.combine_in2pgm_outposterior()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.DataProcessor.form_final_grn"><code class="docutils literal notranslate"><span class="pre">DataProcessor.form_final_grn()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.FileNotFound"><code class="docutils literal notranslate"><span class="pre">FileNotFound</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.ParameterExtractor"><code class="docutils literal notranslate"><span class="pre">ParameterExtractor</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.ParameterExtractor.extract"><code class="docutils literal notranslate"><span class="pre">ParameterExtractor.extract()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.PosteriorCombiner"><code class="docutils literal notranslate"><span class="pre">PosteriorCombiner</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.PosteriorCombiner.combine"><code class="docutils literal notranslate"><span class="pre">PosteriorCombiner.combine()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.Step3ArgumentParser"><code class="docutils literal notranslate"><span class="pre">Step3ArgumentParser</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.Step3ArgumentParser.parse"><code class="docutils literal notranslate"><span class="pre">Step3ArgumentParser.parse()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.Utils"><code class="docutils literal notranslate"><span class="pre">Utils</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.Utils.check_pkl_file_count"><code class="docutils literal notranslate"><span class="pre">Utils.check_pkl_file_count()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.Utils.configure_logging"><code class="docutils literal notranslate"><span class="pre">Utils.configure_logging()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.Utils.print_python_details"><code class="docutils literal notranslate"><span class="pre">Utils.print_python_details()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#InPheC_Step3.step3_main"><code class="docutils literal notranslate"><span class="pre">step3_main()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#faqs">FAQs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#support-and-contribution">Support and Contribution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#getting-support">Getting Support</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#contributing-to-inpherno-chip">Contributing to InPheRNo-ChIP</a></li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div class="tocsection sourcelink">
    <a href="_sources/usage.rst.txt">
      <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
  </div>
</div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">
  <p class="copyright">
    
      © Copyright 2024, Chen Su, Amin Emad.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">
  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item"><p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.14.3.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>